### PR TITLE
Provide true-async and true-sync versions of GetTextChanges

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Pointed out by jason in: https://github.com/dotnet/roslyn/pull/76491

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 1]
```diff
@@ -416,16 +416,28 @@ public Document WithFilePath(string? filePath)
...
+    /// <summary>
+    /// Similar to <see cref="GetTextChangesAsync(Document, CancellationToken)"/>, but should be used when in a forced
+    /// synchronous context.
+    /// </summary>
     internal ImmutableArray<TextChange> GetTextChangesSynchronously(
         Document oldDocument, CancellationToken cancellationToken)
+    {
+        // Should always complete synchronously since we passed in 'useAsync: false'
+        var result = GetTextChangesAsync(useAsync: false, oldDocument, cancellationToken);
+        return result.VerifyCompleted();
```
Comment (JoeRobich) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs:436]: Neat! I did not know about this.

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs [hunk 2]
```diff
@@ -416,16 +416,28 @@ public Document WithFilePath(string? filePath)
...
-    /// Get the text changes between this document and a prior version of the same document.
-    /// The changes, when applied to the text of the old document, will produce the text of the current document.
+    /// Get the text changes between this document and a prior version of the same document. The changes, when applied
+    /// to the text of the old document, will produce the text of the current document.
     /// </summary>
-    public Task<IEnumerable<TextChange>> GetTextChangesAsync(Document oldDocument, CancellationToken cancellationToken = default)
+    public async Task<IEnumerable<TextChange>> GetTextChangesAsync(Document oldDocument, CancellationToken cancellationToken = default)
     {
-        return Task.FromResult<IEnumerable<TextChange>>(GetTextChangesSynchronously(oldDocument, cancellationToken));
+        return await GetTextChangesAsync(useAsync: true, oldDocument, cancellationToken).ConfigureAwait(false);
```
Comment (JoeRobich) [inline src/Workspaces/Core/Portable/Workspace/Solution/Document.cs:424]: What is the benefit of awaiting this task here instead of just returning it?

**Review by JoeRobich — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/Document.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/Document.cs:424]: one returns an ImmutableArray, the other an IEnumerable.  We can't change the IEnumerable (part of hte public API).  but it means callers to the internal methods don't deal with the confusing, potentially lazy, call signature.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Awesome, thanks!

**Pull request merged**